### ISSUE TITLE
Question about weight decay

### ISSUE THREAD

**Issue by davsolai** [ultralytics/yolov5#4104]
## Additional context

Thank you for your contributions.

I have a question about weight decay.

In train.py, 

```
 for k, v in model.named_modules():
        if hasattr(v, 'bias') and isinstance(v.bias, nn.Parameter):
            pg2.append(v.bias)  # biases
        if isinstance(v, nn.BatchNorm2d):
            pg0.append(v.weight)  # no decay
        elif hasattr(v, 'weight') and isinstance(v.weight, nn.Parameter):
            pg1.append(v.weight)  # apply decay
...
if opt.adam:
        optimizer = optim.Adam(pg0, lr=hyp['lr0'], betas=(hyp['momentum'], 0.999))  # adjust beta1 to momentum
    else:
        optimizer = optim.SGD(pg0, lr=hyp['lr0'], momentum=hyp['momentum'], nesterov=True)

    optimizer.add_param_group({'params': pg1, 'weight_decay': hyp['weight_decay']})  # add pg1 with weight_decay
    optimizer.add_param_group({'params': pg2})  # add pg2 (biases)
```

there are different settings for parameters.

Only 'learning rate' is applied without 'weight_decay'.

What is a reason for this setting?
Are there paper or ref for this issue?

Thank you.

**Comment by glenn-jocher** [comment]
@SoYeol-Lee these is the YOLOv5 parameter groupings and decay schedules. Bias LR is initialized high to allow for fast improvements in bias during warmup.

**Comment by vishnubanna** [comment]
is weight decay applied to beta and gamma in the batch normalization layer

**Comment by glenn-jocher** [comment]
@vishnubanna no

**Comment by ashwin-999** [comment]
> @vishnubanna no

Hi @glenn-jocher understand it's based on settings from the paper. Going thru some older issues, I see you've done quite a lot of ablation studies with yolov5. Do you have any intuition of why this specific choice was made? Very curious

Edit: found this link that sorta tries to explain. https://stats.stackexchange.com/questions/576463/why-not-perform-weight-decay-on-layernorm-embedding. Do you see it the same way?

**Comment by glenn-jocher** [comment]
@ashwin-999, the decision not to apply weight decay to the parameters of batch normalization layers (specifically, the beta and gamma parameters) is indeed rooted in empirical findings and theoretical considerations. The primary intuition behind this choice is that these parameters are part of the normalization process, which aims to stabilize the distribution of activations throughout the network. Applying weight decay to these could potentially destabilize these distributions, leading to less effective training.

The link you found provides a good discussion on a related topic, focusing on why weight decay is not typically applied to normalization layers and embeddings. The reasoning aligns with our approach: normalization layers (like BatchNorm) and embeddings have a different role compared to typical weight matrices in a network. They are not directly involved in the "learning" of feature representations in the same way but are more about scaling and shifting those representations. Applying weight decay could hinder their ability to effectively scale and shift, which is why it's generally avoided.

In summary, the choice is based on both theoretical considerations and empirical evidence suggesting that excluding these parameters from weight decay leads to better training dynamics and model performance.
